### DISCUSSION TITLE
Counter Widget Rendering

### DISCUSSION THREAD

**Discussion by eradman** [getredash/redash#6542] [General]
The Counter visualization is relatively simple, but the method it uses to format text is very imporant since it often a prominent part of a dashboard.

In Redash 10, the counter was scalled using `transform()` to fit within the height and widge of the visible space. This caused text to wrap even if there was ample free horizontal space. To solve this we merged a change that modifies the `font-size` base on the `clientHeight` of the parent container.

This works very well for some use cases, but we need to think more carefully since the counter is used to display different kinds of information:

* Visuallization may contain 1, 2, or 3 lines of text (_Label/Value/Target Value_)
* Values may be a short number or more descriptive text

Any individual change we make will likely improve the display for some, and worse for others, so there probably isn't a single _correct_ method. Some considerations:

1. Allow user to select scalling method? (_width/height/area_ of parent container)
2. Allow user to bias the scaling method with a "sale factor"
3. Allow user to specify a fixed font size?

We also need to test the rendering using a HiDPI display (`window.devicePixelRatio=2`)

Below are some screen shots before and after recent commits:

----

Redash 10.1.0:

![redash10 1 0-counters](https://github.com/getredash/redash/assets/1175398/ca44252b-f964-49fe-b977-57ff588e5f5e)

After commit 2795e1b7

![redash-2795e1b7-counters](https://github.com/getredash/redash/assets/1175398/aec8d6c5-c4c9-422a-bf82-9e5a6322ac54)

After commit 593b6ae

![redash-593b6ae-counters](https://github.com/getredash/redash/assets/1175398/099dd2a1-fe6c-4cb9-bd24-0e07a34992f0)

**Comment by justinclift** [comment]
Well, what is the end result (for users) when it's working optimally?

Maybe something like "2/3 of width and height taken up"?

If we can figure out the goal we're after, we should be able to think that backwards through how to implement it.

**Comment by justinclift** [comment]
@myonlylonely This discussion will probably be relevant to you. :smile:

**Reply by myonlylonely** [reply]
The text size in the dashboard and embed queries is still very small. no matter on a HiDPI or a non-HiDPI screen.
This is a embed query, The size is not correct on both HiDPI or a non-HiDPI screen. So this problem may not related to HiDPI at all, but something else.
![image](https://github.com/getredash/redash/assets/8188177/591825ec-fad3-48ef-8a49-1ddd0fa8f37d)

**Comment by justinclift** [comment]
Is there some way for us to get the available space (vertical, horizontal) inside the widget as a first step?

With that, we could probably then try and figure out how much of that space we want to use (maybe 2/3 or 3/4 of whichever dimension is largest?), then work out how big the font needs to be for that to happen.

Line wrapping and/or line breaks could make for an interesting challenge though.

**Comment by eradman** [comment]
One of the challenges is this: in some conitions there is no vertical space alotted to the visualization container:

* Tight horizontal width (mobile view)
* Visualization preview
* Embeded query

The chart visualization asserts it's own minimum height:

```
.chart-visualization-container {
  height: 400px;
  overflow: hidden;
}
```

In Redash 10 the counter font size was fixed at 80px so that it would make room for itself. Here is what the DOM looks like in mobile view:

![redash10-counter-divs](https://github.com/getredash/redash/assets/1175398/0e323b43-8bba-4303-b812-f804c9be1659)

If we follow the same pattern for the counter we should set a minimum height and _then_ we can figure out the right math for the font size.

**Comment by eradman** [comment]
Based on the feedback on https://github.com/getredash/redash/pull/6552 I think the mistake I'm making with font scaling is that the relationship between client height and font size should not be linear.

My first formula:

```sql
SELECT least(80, 8 + (size / 8)) AS "fontSize", size AS "clientHeight"
FROM generate_series(0, 800, 20) AS size
```
![font-scale-linear](https://github.com/getredash/redash/assets/1175398/ef754202-3ab7-4ccb-a2e2-5f353b212f2f)

But to somewhat mimic what the transform() was accomplishing in Redash 10 it should be a power factor

```sql
SELECT 6 + (2.5 * pow(size, 0.5)) AS "fontSize", size AS "clientHeight"
FROM generate_series(1, 800, 20) AS size

```

![font-scale-power-05](https://github.com/getredash/redash/assets/1175398/59974868-19e0-40f5-acdf-dafba221625d)
